ESPS/general/src/esig2fea/esig2fea.1
.\" copyright (c) 1995 Entropic Research Laboratory, Inc. All rights reserved.
.\" @(#)esig2fea.1	1.5 10/4/97 ERL
.ds ]W (c) 1993 Entropic Research Laboratory, Inc.
.if t .ds - \-\h'-0.2m'\-
.if n .ds - ---
.TH ESIG2FEA 1\-ESPS 10/4/97
.SH NAME
.nf
esig2fea \- write Esignal file data to ESPS FEA file
.fi
.SH SYNOPSIS
.B esig2fea
[
.BI \-f " field"
] .\ .\ . [
.BI \-r " range"
] [
.BI \-x " debug_level"
] [
.BI \-F
] [
.BI \-T " subtype"
]
.I input.esig output.fea
.SH DESCRIPTION
.PP
.I esig2fea
accepts an input file
.I input.esig
in Esignal format (see Ref. [1]) and writes an output ESPS FEA file
.I output.fea
(see
.IR ESPS (5\-ESPS),
.IR FEA (5\-ESPS))
containing the same information
(or a subest thereof).
If
.I input.esig
is "\-", standard input is used for the input file.
If
.I output.fea
is "\-", standard output is used for the output file.
The input and output should not be the same file;
however, you can run the program as a filter
by specifying "\-" for both input and output.
.PP
Normally the program reads all input records and copies to the output
the contents of all fields in each record that meet certain requirements;
one output record is written for each input record.
However, with the
.B \-r
option, one can restrict processing to a subrange of the input records,
and with the
.B \-f
option, one can restrict processing to a specified subset of the input fields.
.PP
The requirements on fields
to be copied from input records to output records are the following.
.TP
(1)
Each such field must be a REQUIRED field.  OPTIONAL fields are ignored.
.TP
(2)
Each such field must be a top-level field\*-not a subfield of some
other field.  This means that its name does not contain a dot (".").
Subfields are ignored.
.TP
(3)
Each such field must have a data type that corresponds to an ESPS data
type according to the following table.
.nf
.RS
.ta 0.1i 1.3i 2.6i 4.3i

.if n \l'2.2i\-'\h'0.3i'\l'3.0i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'3.0i'\v'0.2'
	Esignal	ESPS	Esignal     	ESPS
.if n \l'2.2i\-'\h'0.3i'\l'3.0i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'3.0i'\v'0.2'
	DOUBLE	DOUBLE	DOUBLE_COMPLEX	DOUBLE_CPLX
	FLOAT	FLOAT	FLOAT_COMPLEX	FLOAT_CPLX
	LONG	LONG	LONG_COMPLEX	LONG_CPLX
	SHORT	SHORT	SHORT_COMPLEX	SHORT_CPLX
	SCHAR	BYTE	SCHAR_COMPLEX	BYTE_CPLX
	UCHAR	CHAR
	CHAR	CHAR
.if n \l'2.2i\-'\h'0.3i'\l'3.0i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'3.0i'\v'0.2'

.DT
.RE
.fi
Fields of types ARRAY, USHORT, ULONG, BOOL, and WCHAR
are ignored.
.LP
Fields that violate these requirements are allowed in input files;
however,
.I esig2fea
does not provide a mapping of such fields to FEA fields.
Each input field that satisfies the requirements (and is retained)
normally becomes an output field with the same name.
.PP
An exception is a REQUIRED field with data type LONG
and rank 0 (a scalar).
If such a field is present (and is retained), the output file is
tagged (see
.IR ESPS (5\-ESPS)),
and the "Tag" value in each Esignal record
furnishes the value for the tag in the corresponding FEA record,
rather than for an ordinary FEA field named "Tag".
.PP
Each top-level GLOBAL field with one of the Esignal types shown
in the following table
.nf
.RS
.ta 0.1i 1.3i 2.6i 3.8i

.if n \l'2.2i\-'\h'0.3i'\l'2.2i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'2.2i'\v'0.2'
	Esignal	ESPS	Esignal	ESPS
.if n \l'2.2i\-'\h'0.3i'\l'2.2i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'2.2i'\v'0.2'
	DOUBLE	DOUBLE	SCHAR	CHAR
	FLOAT	FLOAT	UCHAR	CHAR
	LONG	LONG	CHAR	CHAR
	SHORT	SHORT
.if n \l'2.2i\-'\h'0.3i'\l'2.2i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'2.2i'\v'0.2'

.DT
.RE
.fi
becomes an ESPS generic header item with the same name,
the indicated corresponding ESPS type,
and the same number of data elements.
ESPS generic header items
do not contain rank and dimension information, however.
Thus for example a 15-element vector and a 3\(mu5 matrix in
Esignal GLOBAL fields both become simply 15-element generic header items.
.PP
If an Esignal GLOBAL or REQUIRED field with Esignal data type SHORT
has a subfield named "enumStrings",
the corresponding FEA file field or header item
has an ESPS CODED type, (see
.IR FEA (5\-ESPS))
rather than the normal corresponding ESPS type SHORT
indicated in the table.
The subfield must be a character matrix consisting of strings,
one per row, padded on the right with null characters ('\\0') to a
common length.
The strings become the list of string values associated with
the CODED type.
.PP
The output FEA file header has a single embedded source file header
(see
.IR ESPS (5\-ESPS),
.IR add_source_file (3\-ESPS))
which, like the output file header itself, is derived from the
field list in the input Esignal header.
The source header is thus usually similer to the output file header,
but there are some differences.
The source header records all the fields of the input file
that satisfy (1), (2), and (3) above, even if the
.B \-f
option is used to specify that only some of these fields
are retained in the output.
Within the source header,
there is a recursively embedded source header
for each VIRTUAL field (if any) with a name of the form
"source_0", "source_1", "source_2", etc.
The embedded source header is obtained by converting the subfield list
of the VIRTUAL field from an Esignal field list to an ESPS FEA file header.
If the VIRTUAL field has a data array,
it must be a one-dimensional CHAR array giving the source-file name
that goes with the corresponding header.
More deeply nested source-file headers are derived from VIRTUAL
subfields with names such as "source_0.source_2"
or "source_0.source_2.source_1".
.PP
If the input file has a scalar GLOBAL field
.I FeaSubtype
of type SHORT,
.I esig2fea
assigns its value to the FEA file subtype code
.I fea_type
(see
.IR FEA (5\-ESPS)).
This behavior can be overridden by the
.B \-T
option.
.PP
If the input file contains the waves GLOBAL fields
.I recordFreq
and
.I startTime,
the program creates corresponding waves generic header items
.I record_freq
and
.I start_time
in the output file.
These are the header items that
.IR xwaves (1\-ESPS)
uses for time alignment of displays of files with uniformly spaced records.
The value of
.I record_freq
is just that of
.I recordFreq.
The value of
.I start_time
is obtained from that of
.I startTime
by adding an offset that is the time difference
between the first record of the input file
and the first record that is copied to the output.
The offset is 0 unless the
.B \-r
option is used to specify a starting point other than the first input record.
(See the ``Options'' section.)
In that case, the offset is the result of dividing the starting record number
by the value of
.I recordFreq.
Like other input GLOBAL fields,
.I recordFreq
and
.I startTime,
if present, are represented in the output
by generic header items with the same names.
In this case, the header item
.I startTime
is simply a copy of the input GLOBAL field
.I startTime;
it is the header item
.I start_time
that contains the correct starting time of the output file.
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-f " field"
specifies the name of a input field
to be retained in the output file records.
The field should be a top-level field (the name should not contain "."),
it should be a REQUIRED field (not OPTIONAL, for example),
and it should have a supported data type (not ARRAY, for example);
otherwise it is ignored.
The option may be repeated to specify several input fields to be retained.
If the option is not specified,
all top-level REQUIRED input fields of supported types are retained.
.TP
.BI \-r " start" : "last \fR[0:(last in file)]\fP"
.TP
.BI \-r " start" :+ "incr"
.TP
.BI \-r " start"
In the first form, a pair of unsigned integers specifies the range
of input records to be processed.
Input record numbers follow Esignal file conventions:
the first record in the file is number 0.
Either
.I start
or
.I last
may be omitted; then the default value is used.
If
.IR last " = " start " + " incr,
the second form (with the plus sign) specifies the same range as the first.
The third form (omitting the colon) specifies a single record.
.TP
.BI \-x " debug_level \fR[0]\fP"
If 
.I debug_level
is positive,
.I fea2esig
writes debugging messages on the standard error output.
Larger values result in more output.
The default is 0, for no output.
.TP
.BI \-F
specifies output in
.I field order.
If the option is not specified, the output is in
.I type order,
the default.
.TP
.BI \-T " subtype"
specifies that the subtype code
.I fea_type
of the output FEA file (see
.IR FEA (5\-ESPS))
is filled in to indicate the given subtype.
Allowed values include FEA_ANA, FEA_SD, FEA_SPEC, FEA_STAT,
and FEA_VQ.
The complete list of subtype names,
with the corresponding numeric codes,
is in the ESPS header file
.I esps/fea.h.
In addition,
.I subtype
may be NONE,
to specify that the output is not of any special FEA subtype.
If
.B \-T
is not specified,
.I esig2fea
looks in the input file for a GLOBAL field named "FeaSubtype"
and having data type SHORT.
It sets the subtype code according to the value of
.I FeaSubtype
if the field is present;
otherwise it sets the code to 0, for NONE.
The
.B \-T
option overrides any values specified by an input field
.I FeaSubtype.
.IP
Such a field
.I FeaSubtype
is created when the program
.IR fea2esig (1\-ESPS)
converts a file of a special FEA subtype to
.I Esignal
format.
This is done only so that
.I esig2fea
can restore the subtype when converting the file back to a FEA file;
the notion of a FEA subtype has no exact parallel in the Esignal
framework.
.IP
Use of
.B \-T
doesn't ensure that the output file has the correct format.
You should specify a subtype with this option only when you know
that a file with the correct fields and generic header items
will be produced.
See the section-5 manual pages for the individual FEA file types.
.SH ESPS PARAMETERS
.PP
The ESPS parameter file is not read.
.SH ESPS COMMON
.PP
The ESPS common file is not accessed.
.SH ESPS HEADERS
.PP
As mentioned under ``Description'' above,
generic header items corresponding to top-level GLOBAL fields
are added to the output header,
and the input field list, converted to a FEA header,
is added as an embedded source header.
As is usual in ESPS files, the command line that invoked the program
is added as a comment.
For the setting of the FEA subtype code
.I fea_type,
see
.B \-T
under ``Options''.
Generic header items
.I record_freq
and
.I start_time
are added to the output header if the input contains GLOBAL fields
.I recordFreq
and
.I startTime;
see ``Description'' above for details.
.SH FUTURE CHANGES
.PP
Fix some of the problems listed under ``Bugs''.
.SH EXAMPLES
.PP
.SH ERRORS AND DIAGNOSTICS
.PP
The program exits with a "Usage" message
if an unrecognized option is used on the command line.
or if too many or too few file names are specified.
It may exit with one of the following error messages
if the described conditions occur:
.nf
.RS
can't start before beginning of file
empty range of records specified
can't allocate memory for input record
failure converting input header to field list
output order neither TYPE_ORDER nor FIELD_ORDER
can't set field ordering of output
write header failed
.RE
.fi
.SH BUGS
.PP
Confusion is possible if field names in the input file
conflict with field names that
.I esig2fea
uses for special purposes\*-for example,
if there is a REQUIRED input field named "commandLine" or "source_0".
The program ignores subfields named "encoding"
(which flag their parent fields as containing
.if n .IR mu -law
.if t .IR \(*m -law
or
.IR A -law
compressed data.
.SH REFERENCES
.PP
[1] R. W. Johnson,
``The Esignal File Format'',
Entropic Research Laboratory, Inc., 1995
.SH "SEE ALSO"
.IR add_source_file (3\-ESPS),
.IR fea2esig (1\-ESPS),
.IR xwaves (1\-ESPS),
.br
.IR Esignal (3\-Esignal),
.IR ESPS (5\-ESPS),
.IR FEA (5\-ESPS).
.PP
.SH AUTHOR
.PP
Rod Johnson


ESPS/general/src/esig2fea/esig2fea.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1995  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Rod Johnson
 * Checked by:
 * Revised by:
 *
 * Write Esignal file data to FEA file.
 *
 */

static char *sccs_id = "@(#)esig2fea.c	1.4	2/5/97	ERL";

#include <esps/esps.h>
#include <esps/fea.h>
/* NAME CONFLICT:  The ESPS data-type codes named
 * DOUBLE, FLOAT, LONG, SHORT, and CHAR
 * are different from the Esignal data-type codes with the same names.
 * While the ESPS definitions are in effect, declare variables so that
 * the ESPS codes can be referred to as
 * FDOUBLE, FFLOAT, FLONG, FSHORT, and FCHAR
 * instead.  (Think of "fea-double", "fea-float", etc.)
 * After the inclusion of <esignal.h> below, the Esignal definitions of
 * DOUBLE, FLOAT, LONG, SHORT, and CHAR
 * take effect.
 */
const static int FDOUBLE = DOUBLE;
#undef	DOUBLE
const static int FFLOAT = FLOAT;
#undef	FLOAT
const static int FLONG = LONG;
#undef	LONG
const static int FSHORT = SHORT;
#undef	SHORT
const static int FCHAR = CHAR;
#undef	CHAR
#undef	BOOL
#include <esignal.h>

#define SCCS_VERSION	    "1.4"
#define SCCS_DATE	    "2/5/97"

#define ESPS_TAG	    "Tag"
#define ESPS_FEA_SUBTYPE    "FeaSubtype"

#define REQUIRE(test, text) {if (!(test)) {(void) fprintf(stderr, \
"%s: %s - exiting\n", ProgName, text); exit(1);}}

#define SYNTAX \
USAGE("esig2fea [-f field]... [-r range] [-x debug_level] [-F] [-T subtype]\n\tinput.esig output.fea");

void	lrange_switch();
char	*savestring();
char	*get_cmd_line();


static struct header	*FieldList_to_fea(FieldList list,
					  struct fea_data **rec,
					  char **fnames, int copy_sources);
static int		FindStr(char *str, char **arr);
static char		**StrArrayFromRect(long *dim, void *data);
static int		FieldIsTag(FieldSpec *fld);
static int		FieldIsFeaSubtype(FieldSpec *fld);
static long		SkipRecs(FILE *file, long skip,
				 long size, FieldSpec **fields, int arch);
static int		ElibTypeToEsps(int type);


int	    debug_level = 0;


main(int   argc,
     char  **argv)
{
    extern int	    optind;	/* for use of getopt() */
    extern char	    *optarg;	/* for use of getopt() */
    int		    ch;		/* command-line option letter */

    static char	    *ProgName = "esig2fea";	/* name of this program */
    static char	    *Version = SCCS_VERSION;	/* program SCCS version */
    static char	    *Date = SCCS_DATE;		/* program SCCS date */

    char	    **field_names = NULL;
    int		    num_fields = 0;
    int		    alloc_fields = 0;

    int		    rflag = NO;	/* -r option specified? */
    char	    *rrange;	/* arguments of -r option */
    long	    start_rec;	/* starting record number */
    long	    end_rec;	/* ending record number */
    long	    num_recs;	/* number of records to read
				   (0 means all up to end of file) */
    long	    num_read;	/* number of records actually read */

    char	    *iname;	/* input file name */
    FILE	    *ifile;	/* input stream */
    FieldList	    list;	/* input field list */
    int		    inord;	/* input: field order or type order? */
    FieldSpec	    **ifields;	/* input fields in field or type order */

    char	    *subtype = NULL;		/* FEA subtype name */
    int		    subtype_code = NONE;	/* numeric subtype code */
    FieldSpec	    *fld;	/* spec of various special fields */

    char	    *oname;	/* output file name */
    FILE	    *ofile;	/* output stream */
    struct header   *ohd;	/* output file header */
    struct fea_data *orec;	/* output record */
    int		    outord = TYPE_ORDER;

    char	    *version;	/* version from input preamble */
    int		    arch;	/* architecture from input preamble */
    long	    pre_size;	/* preamble size */
    long	    hdr_size;	/* header size (bytes) from preamble */
    long	    rec_size;	/* record size from preamble */

    double	    rec_freq;
    double	    start_time_offset;
    double	    *data;
    long	    len, i;


    struct header   *source;	/* embedded source-file header */


    while ((ch = getopt(argc, argv, "f:r:x:FT:")) != EOF)
	switch (ch)
	{
	case 'f':
	    if (num_fields >= alloc_fields)
	    {
		size_t	size;

		alloc_fields = num_fields + 1 + num_fields/2;
		size = (alloc_fields + 1) * sizeof(char *);
		field_names = (char **)
		    ((field_names == NULL)
		     ? malloc(size)
		     : realloc(field_names, size));
	    }
	    field_names[num_fields++] = optarg;
	    field_names[num_fields] = NULL;
	    break;
	case 'r':
	    rflag = YES;
	    rrange = optarg;
	    break;
	case 'x':
	    debug_level = atoi(optarg);
	    break;
	case 'F':
	    outord = FIELD_ORDER;
	    break;
	case 'T':
	    subtype = optarg;
	    break;
	default:
	    SYNTAX;
	    break;
	}

    if (argc - optind > 2)
    {
	fprintf(stderr,
		"%s: too many file names specified.\n", ProgName);
	SYNTAX;
    }
    if (argc - optind < 2)
    {
	fprintf(stderr,
		"%s: too few file names specified.\n", ProgName);
	SYNTAX;
    }


    DebugMsgLevel = debug_level;
    DebugMsgFunc = DebugPrint;

    iname = argv[optind++];
    list = OpenIn(iname, &version,
		  &arch, &pre_size, &hdr_size, &rec_size, &ifile);
    REQUIRE(list != NULL, "read header failed");
    if (ifile == stdin)
	iname = "<stdin>";

    oname = argv[optind++];

    start_rec = 0;
    end_rec = LONG_MAX;
    num_recs = 0;	/* 0 means continue to end of file */

    if (rflag)
    {
	lrange_switch(rrange, &start_rec, &end_rec, 0);
	if (end_rec != LONG_MAX)
	    num_recs = end_rec - start_rec + 1;
    }

    REQUIRE(start_rec >= 0, "can't start before beginning of file");
    REQUIRE(end_rec >= start_rec, "empty range of records specified");

    if (debug_level)
	fprintf(stderr,
		"start_rec: %ld.  end_rec: %ld.  num_recs: %ld.\n",
		start_rec, end_rec, num_recs);

    REQUIRE(GetFieldOrdering(list, &inord),
	    "cant get field ordering of input");

    switch (inord)
    {
    case TYPE_ORDER:
	if (debug_level)
	    fprintf(stderr, "making type-ordered field array.\n");
	ifields = TypeOrder(list);
	break;
    case FIELD_ORDER:
	if (debug_level)
	    fprintf(stderr, "making field-ordered field array.\n");
	ifields = FieldOrder(list);
	break;
    default:
	REQUIRE(0, "input order neither TYPE_ORDER nor FIELD_ORDER");
	break;
    }

    ohd = FieldList_to_fea(list, &orec, field_names, FALSE);

    REQUIRE(ohd != NULL,
	    "failure converting input field list to header & record struct");

    if (subtype != NULL)
    {
	subtype_code = lin_search(fea_file_type, subtype);

	if (subtype_code == -1)
	    fprintf(stderr, "%s: unknown FEA file subtype \"%s\" ignored.\n",
		    ProgName, subtype);
	else
	    ohd->hd.fea->fea_type = subtype_code;
    }

    if (outord == FIELD_ORDER)
	ohd->hd.fea->field_order = YES;

    fld = FindField(list, "recordFreq");

    if (fld != NULL && fld->occurrence == GLOBAL && fld->data != NULL)
    {
	(void) type_convert(1L, (char *) fld->data, ElibTypeToEsps(fld->type),
			    (char *) &rec_freq, FDOUBLE, (void (*)()) NULL);

	*add_genhd_d("record_freq", NULL, 1, ohd) = rec_freq;
    }
    else
	rec_freq = 1.0;

    fld = FindField(list, "startTime");

    if (fld != NULL
	&& fld->occurrence == GLOBAL && fld->data != NULL && rec_freq != 0)
    {
	start_time_offset = start_rec / rec_freq;

	len = FieldLength(fld);
	data = (double *) type_convert(len, (char *) fld->data,
				       ElibTypeToEsps(fld->type),
				       (char *) NULL, FDOUBLE,
				       (void (*)()) NULL);

	if (start_time_offset != 0)
	{
	    for (i = 0; i < len; i++)
		data[i] += start_time_offset;
	}

	(void) add_genhd_d("start_time", data, len, ohd);
    }

    (void) strcpy(ohd->common.prog, ProgName);
    (void) strcpy(ohd->common.vers, Version);
    (void) strcpy(ohd->common.progdate, Date);

    source = FieldList_to_fea(list, NULL, NULL, TRUE);
    add_source_file(ohd, savestring(iname), source);

    add_comment(ohd, get_cmd_line(argc, argv));

    oname = eopen(ProgName,
		  oname, "w", NONE, NONE, NULL, &ofile);
    write_header(ohd, ofile);

    num_read = SkipRecs(ifile, start_rec,
			RecordSize(list, arch), ifields, arch);

    if (num_read != start_rec)
    {
	fprintf(stderr,
		"%s: couldn't reach starting record; only %ld skipped.\n",
		ProgName, num_read);
	exit(0);
    }

    for ( ;
	 num_read <= end_rec && ReadRecord(ifields, arch, ifile);
	 num_read++)
    {
	put_fea_rec(orec, ohd, ofile);
    }

    if (num_read <= end_rec && num_recs != 0)
	fprintf(stderr, "esig2fea: only %ld records read.\n",
		num_read - start_rec);

    exit(0);
    /*NOTREACHED*/
}


static struct header *
FieldList_to_fea(FieldList	    list,
		 struct fea_data    **rec,
		 char		    **fnames,
		 int		    copy_sources)
{
    struct header	*hdr;
    int			i, j;
    FieldSpec		*fld;
    char		**codes;
    FieldList		subfields;
    FieldSpec		*subf;
    char		*name;
    void		*ptr;
    int			spstype;
    long		*dim;
    struct fea_header	*fea;
    struct header	*source;

    if (list == NULL)
	return NULL;

    hdr = new_header(FT_FEA);

    for (i = 0; (fld = list[i]) != NULL; i++)
    {
	codes = NULL;
	if (fld->occurrence != VIRTUAL
	    && (subfields = fld->subfields) != NULL)
	{
	    for (j = 0; (subf = subfields[j]) != NULL; j++)
	    {
		if (strcmp(subf->name, "enumStrings") != 0)
		{
		    if (debug_level)
			fprintf(stderr,
				"FieldList_to_fea: subfields "
				"not supported in ESPS FEA files.\n");
		}
		else if (fld->type != SHORT)
		{
		    if (debug_level)
			fprintf(stderr,
				"FieldList_to_fea: Non-SHORT field "
				"has subfield enumStrings.\n");
		}
		else if (subf->type != CHAR)
		{
		    if (debug_level)
			fprintf(stderr,
				"FieldList_to_fea: enumStrings not CHAR.\n");
		}
		else if (subf->rank != 2)
		{
		    if (debug_level)
			fprintf(stderr,
				"FieldList_to_fea: enumStrings "
				"not of rank 2.\n");
		}
		else
		    codes = StrArrayFromRect(subf->dim, subf->data);
	    }
	}

	if (FieldIsTag(fld) && FindStr(ESPS_TAG, fnames))
	    hdr->common.tag = TRUE;
	else if (FieldIsFeaSubtype(fld))
	    hdr->hd.fea->fea_type = *(short *) fld->data;
	else
	{
	    name = fld->name;

	    switch(fld->occurrence)
	    {
	    case GLOBAL:
		{
		    int     size = (int) FieldLength(fld);

		    ptr = fld->data;

		    if (debug_level >= 2)
			fprintf(stderr,
				"FieldList_to_fea: "
				"global field[%d]: \"%s\".\n",
				i, name);

		    if (fld->rank > 1)
		    {
			if (debug_level)
			    fprintf(stderr,
				    "FieldList_to_fea: rank %d globals "
				    "not supported in ESPS FEA files.\n",
				    fld->rank);
		    }

		    if (size == 0)
		    {
			if (debug_level)
			    fprintf(stderr,
				    "FieldList_to_fea: empty globals "
				    "not supported in ESPS FEA files.\n");
		    }
		    else if (codes != NULL)
			(void) add_genhd_e(name,
					   (short *) ptr, size, codes, hdr);
		    else
			switch (fld->type)
			{
			case DOUBLE:
			    (void) add_genhd_d(name,
					       (double *) ptr, size, hdr);
			    break;
			case FLOAT:
			    (void) add_genhd_f(name,
					       (float *) ptr, size, hdr);
			    break;
			case LONG:
			    (void) add_genhd_l(name,
					       (long *) ptr, size, hdr);
			    break;
			case SHORT:
			    (void) add_genhd_s(name,
					       (short *) ptr, size, hdr);
			    break;
			case SCHAR:
			case UCHAR:
			case CHAR:
			    (void) add_genhd_c(name,
					       (char *) ptr, size, hdr);
			    break;
			default:
			    if (debug_level)
				fprintf(stderr,
					"FieldList_to_fea: global type %d "
					"not supported in ESPS FEA files.\n",
					fld->type);
			}
		}
		break;
	    case REQUIRED:
		{
		    long    size = FieldLength(fld);

		    if (debug_level >= 2)
			fprintf(stderr,
				"FieldList_to_fea: "
				"required field[%d]: \"%s\".\n",
				i, name);

		    if (FindStr(name, fnames))
		    {
			spstype = (codes != NULL) ? CODED
			    : ElibTypeToEsps(fld->type);
			if (spstype != UNDEF)
			{
			    dim = (long *) malloc(fld->rank * sizeof(long));
			    for (j = 0; j < fld->rank; j++)
				dim[j] = fld->dim[j];
			    add_fea_fld(name, size, fld->rank,
					dim, spstype, codes, hdr);
			}
		    }
		}
		break;
	    case OPTIONAL:
		if (debug_level)
		    fprintf(stderr,
			    "FieldList_to_fea: optional fields "
			    "not supported in ESPS FEA files.\n");
		break;
	    case VIRTUAL:
		if (copy_sources)
		{
		    if (strncmp(name, "source_", 7) != 0)
		    {
			if (debug_level)
			    fprintf(stderr, "Field_List_to_fea: VIRTUAL "
				    "field other than source_<n>.\n");
		    }
		    else if ((subfields = fld->subfields) != NULL
			     || fld->type == CHAR)
		    {
			size_t	len;
			char	*data;

			source = FieldList_to_fea(subfields,
						  NULL, NULL, TRUE);
			len = FieldLength(fld);
			data = (char *) malloc(len + 1);
			strncpy(data, fld->data, len);
			data[len] = '\0';

			add_source_file(hdr, data, source);
		    }
		}
		break;
	    case INCLUDED:
		if (debug_level)
		    fprintf(stderr,
			    "FieldList_to_fea: included fields "
			    "not supported in ESPS FEA files.\n");
		break;
	    default:
		spsassert(0, "FieldList_to_fea: "
			  "unrecognized occurrence code.\n");
		break;
	    }
	}
    }

    if (rec != NULL)
    {
	*rec = allo_fea_rec(hdr);
	fea = hdr->hd.fea;

	if (hdr->common.tag)
	{
	    fld = FindField(list, ESPS_TAG);
	    fld->data = &(*rec)->tag;
	}

	for (i = 0; i < (int) fea->field_count; i++)
	{
	    name = fea->names[i];
	    fld = FindField(list, name);
	    fld->data = get_fea_ptr(*rec, name, hdr);
	}
    }

    return hdr;
}


/*!*//* Same function FindStr defined in fea2esig.c. */

static int
FindStr(char	*str,
	char	**arr)
{
    int     i;

    spsassert(str != NULL, "FindStr: NULL string.");

    if (arr == NULL)
	return TRUE;		/* NULL array implies all fields */

    for (i = 0; arr[i] != NULL; i++)
	if (strcmp(str, arr[i]) == 0)
	    return TRUE;

    return FALSE;
}


static char **
StrArrayFromRect(long	*dim,
		 void	*data)
{
    char	**strarr;
    long	len, wid, i;
    char	*row;

    len = 1 + dim[0];

    strarr = (char **) malloc(len * sizeof(char *));
    len -= 1;

    wid = dim[1];
    row = data;
    for (i = 0; i < len; i++)
    {
/*!*//* Should check for null termination of rows. */
	strarr[i] = savestring(row);
	row += wid;
    }

    strarr[len] = NULL;

    return strarr;
}


static int
FieldIsTag(FieldSpec *fld)
{
    if (strcmp(fld->name, ESPS_TAG) != 0)
	return FALSE;
    if (fld->occurrence != REQUIRED)
    {
	if (debug_level)
	    fprintf(stderr,
		    "FieldIsTag: non-REQUIRED field named \"%s\".\n",
		    ESPS_TAG);
	return FALSE;
    }
    if (fld->type != LONG)
    {
	if (debug_level)
	    fprintf(stderr,
		    "FieldIsTag: type of field named \"%s\" not LONG.\n",
		    ESPS_TAG);
	return FALSE;
    }
    if (fld->rank != 0)
    {
	if (debug_level)
	    fprintf(stderr,
		    "FieldIsTag: rank of field named \"%s\" not 0.\n",
		    ESPS_TAG);
	return FALSE;
    }
    return TRUE;
}


static int
FieldIsFeaSubtype(FieldSpec *fld)
{
    if (strcmp(fld->name, ESPS_FEA_SUBTYPE) != 0)
	return FALSE;
    if (fld->occurrence != GLOBAL)
    {
	if (debug_level)
	    fprintf(stderr,
		    "FieldIsFeaSubtype: non-GLOBAL field named \"%s\".\n",
		    ESPS_FEA_SUBTYPE);
	return FALSE;
    }
    if (fld->type != SHORT)
    {
	if (debug_level)
	    fprintf(stderr,
		    "FieldIsFeaSubtype: "
		    "type of field named \"%s\" not SHORT.\n",
		    ESPS_FEA_SUBTYPE);
	return FALSE;
    }
    if (fld->rank != 0)
    {
	if (debug_level)
	    fprintf(stderr,
		    "FieldIsFeaSubtype: rank of field named \"%s\" not 0.\n",
		    ESPS_FEA_SUBTYPE);
	return FALSE;
    }
    if (fld->data == NULL)
    {
	if (debug_level)
	    fprintf(stderr,
		    "FieldIsFeaSubtype: "
		    "field named \"%s\" has NULL data pointer.\n",
		    ESPS_FEA_SUBTYPE);
	return FALSE;
    }
    return TRUE;
}


static long
SkipRecs(FILE	    *file,
	 long	    skip,
	 long	    size,
	 FieldSpec  **fields,
	 int	    arch)
{
    long    n;

    if (skip == 0)
	return 0;

    if (size != -1)
    {
	skiprec(file, skip, size);
	return skip;
    }
    else
    {
	for (n = 0; n < skip && ReadRecord(fields, arch, file); n++)
	{ }

	return n;
    }
}


static int
ElibTypeToEsps(int type)
{
    switch (type)
    {
    case DOUBLE:
	return FDOUBLE;
    case FLOAT:
	return FFLOAT;
    case LONG:
	return FLONG;
    case SHORT:
	return FSHORT;
    case SCHAR:
	return BYTE;
    case UCHAR:
    case CHAR:
	return FCHAR;
    case DOUBLE_COMPLEX:
	return DOUBLE_CPLX;
    case FLOAT_COMPLEX:
	return FLOAT_CPLX;
    case LONG_COMPLEX:
	return LONG_CPLX;
    case SHORT_COMPLEX:
	return SHORT_CPLX;
    case SCHAR_COMPLEX:
	return BYTE_CPLX;
    default:
	return UNDEF;
    }
}
